bundles/org.eclipse.osgi/supplement/src/org/eclipse/osgi/util/ResourceFormatter.java

```diff
@@ -0,0 +1,127 @@
+ * https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *     Christoph Läubrich - extracted from ModuleResolutionReport / ModuleRevision
+ *******************************************************************************/
+package org.eclipse.osgi.util;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Set;
+import org.eclipse.osgi.internal.framework.FilterImpl;
+import org.osgi.framework.Constants;
+import org.osgi.framework.InvalidSyntaxException;
+import org.osgi.framework.namespace.BundleNamespace;
+import org.osgi.framework.namespace.HostNamespace;
+import org.osgi.framework.namespace.PackageNamespace;
+import org.osgi.resource.Capability;
+import org.osgi.resource.Namespace;
+import org.osgi.resource.Requirement;
+
+/**
+ * Utility class that provides some generic Resource API > String conversion
+ * that are suitable to be presented to the user.
+ * 
+ * @since 3.19
+ */
+public final class ResourceFormatter {
+	public static String printCapability(Capability cap) {
+		if (PackageNamespace.PACKAGE_NAMESPACE.equals(cap.getNamespace())) {
+			return Constants.EXPORT_PACKAGE + ": " + createOSGiCapability(cap); //$NON-NLS-1$
+		} else if (BundleNamespace.BUNDLE_NAMESPACE.equals(cap.getNamespace())) {
+			return Constants.BUNDLE_SYMBOLICNAME + ": " + createOSGiCapability(cap); //$NON-NLS-1$
+		} else if (HostNamespace.HOST_NAMESPACE.equals(cap.getNamespace())) {
+			return Constants.BUNDLE_SYMBOLICNAME + ": " + createOSGiCapability(cap); //$NON-NLS-1$
+		}
+		return Constants.PROVIDE_CAPABILITY + ": " + cap.toString(); //$NON-NLS-1$
+	}
+
+	public static String createOSGiCapability(Capability cap) {
+		Map<String, Object> attributes = new HashMap<>(cap.getAttributes());
+		Map<String, String> directives = cap.getDirectives();
+		String name = String.valueOf(attributes.remove(cap.getNamespace()));
+		return name + toString(attributes, false, true) + toString(directives, true, true);
+	}
+
+	public static String printRequirement(Requirement req) {
+		if (PackageNamespace.PACKAGE_NAMESPACE.equals(req.getNamespace())) {
+			return Constants.IMPORT_PACKAGE + ": " //$NON-NLS-1$
+					+ createOSGiRequirement(req, PackageNamespace.CAPABILITY_VERSION_ATTRIBUTE,
+							PackageNamespace.CAPABILITY_BUNDLE_VERSION_ATTRIBUTE);
+		} else if (BundleNamespace.BUNDLE_NAMESPACE.equals(req.getNamespace())) {
+			return Constants.REQUIRE_BUNDLE + ": " //$NON-NLS-1$
+					+ createOSGiRequirement(req, BundleNamespace.CAPABILITY_BUNDLE_VERSION_ATTRIBUTE);
+		} else if (HostNamespace.HOST_NAMESPACE.equals(req.getNamespace())) {
+			return Constants.FRAGMENT_HOST + ": " //$NON-NLS-1$
+					+ createOSGiRequirement(req, HostNamespace.CAPABILITY_BUNDLE_VERSION_ATTRIBUTE);
+		}
+		return Constants.REQUIRE_CAPABILITY + ": " + req.toString(); //$NON-NLS-1$
+	}
+
+	public static String createOSGiRequirement(Requirement requirement, String... versions) {
```

Comment:
Same as `createOSGiCapability`, this should not be public.  It is an internal detail of `printRequirement`